### DISCUSSION TITLE
Custom plugin

### DISCUSSION THREAD

**Discussion by nagarjun1143** [glowroot/glowroot#1255] [Q&A] [answered]
Is it possible to have trace entry and query entry for the same pointcut? How do we make sure that we only capture parameters first time it intercepted method and not the internal calls

**Comment by Never-lab** [comment, marked as answer]
This looks like a **plugin / custom-instrumentation how-to**, not a defect — so the best place for follow-ups is **[GitHub Discussions → Q&A](https://github.com/glowroot/glowroot/discussions/categories/q-a)** (see the [README “Where to ask”](https://github.com/glowroot/glowroot/blob/main/README.md) table). I’ll answer briefly here and we can continue there.

### 1) Trace entry and query entry for the same pointcut?

**Yes — use a `QueryEntry`.** In the plugin API, `QueryEntry` **extends** `TraceEntry`: one `ThreadContext.startQueryEntry(...)` creates a specialized entry that shows up in the trace timeline *and* is aggregated by query text (same pattern as the JDBC plugin).

You normally do **not** call both `startTraceEntry` and `startQueryEntry` for the same span. Prefer:

```java
QueryEntry entry = context.startQueryEntry(queryType, queryText, messageSupplier, timerName);
// … method body …
entry.end();
```

### 2) Capture parameters only on the outer call (skip nested / internal re-entries)?

Also supported:

- **Java plugin:** set a shared `nestingGroup` on `@Pointcut`. While that group is active, nested advice in the same group is suppressed (used heavily by built-in plugins, e.g. servlet `outer-servlet-or-filter`).
- **UI “Instrumentation”:** leave **“Capture self nested trace entries”** unchecked (default). Nested hits under the same timer name won’t create extra entries — useful when overloads cascade into each other.
- **Advanced:** `suppressionKey` / `suppressibleUsingKey` on `@Pointcut` for cross-pointcut suppression.

So both goals are doable with existing APIs; please reopen any remaining detail in [Discussions → Q&A](https://github.com/glowroot/glowroot/discussions/categories/q-a) (feel free to link this issue). Closing here as not a bug.
